src/Microsoft.AspNet.Server.Kestrel/Http/SocketOutput.cs

```diff
@@ -91,6 +92,21 @@ public class SocketOutput : ISocketOutput
                 if (buffer.Count > 0)
                 {
                     var tail = ProducingStart();
+                    if (tail.IsDefault)
+                    {
+                        if (cancellationToken.CanBeCanceled)
+                        {
+                            return TaskUtilities.GetCancelledTask(
+                                cancellationToken.IsCancellationRequested ?
+                                cancellationToken :
+                                new CancellationToken(true));
```

Comment:
This whole flow is riddled with race conditions so you're never going to get a deterministic behavior. The one behavior we do want to guarantee is the following: When I call WriteAsync with a cancellable CT and the operation becomes async, then I should be able to cancel that CT and have the async operation end in a timely fashion. Even here there is a race here where the operation may complete successfully just before being cancelled, or it may throw a TaskCancelledException (or equivalent). We're not trying to grantee it throws, only that it ends. If there's a race then the cancelled CT will still fail next time it's checked and terminate operations.